### ISSUE TITLE
Convert training from full-batch to mini-batch

### ISSUE THREAD

**Issue by finlaymorrison** [finlaymorrison/BunjiLearn#24]
Currently the `Trainer::fit` function only implements full-batch gradient descent. This function should take in an integer which defines the batch size to use when training. This batch size can be set to the dataset size to give full-batch gradient descent. This change will open up the possibility of making the optimiser configurable in a later change.
